### ISSUE TITLE
Vote gaming, channel parting/joining, multiple clients, etc.

### ISSUE THREAD

**Issue by DanielOaks** [KnightHawk3/Rhythm#13]
This is **not** a high priority issue, or something we need to find a firm solution to right away. It's more a discussion point around what we wanna do (if anything) regarding some different sorts of intentional or accidental vote gaming people can do.

There are a few ways people can get two votes, etc.

* User is in the channel. Motion starts, person casts a vote, leaves the channel, joins the channel under a different nick, and then casts another vote.
* User is in the channel. Motion starts, person casts a vote, person brings another one of their clients into the channel and casts another vote on the second client.
* User has two clients in the channel. Motion starts, person casts a vote on first client, person casts a vote on second client.

The easy thing to say is 'votes should be tallied by a human' and ignore it, but there could be some easy ways to prevent people from accidentally casting more than one vote. Perhaps storing the username/host mask when users vote and then cross-checking that instead of / in addition to nicknames?

The problem with implementing a solution to this is that we can really only hope to prevent accidental double-votes. If someone really wants to bypass the controls, they'll find a way with vhosts, hostnames, just getting a VPS/VPN, etc.

**Comment by olligobber** [comment]
Another easy thing to say is to carefully only voice users who can vote. If there's no automatic voicing, and each voter is only voiced under one nick, repeat voting is mostly avoidable. One exception is if a user votes, swaps nicks, disconnects and then rejoins under their original nick. They might be voiced and then vote again accidentally. There is also room for intentional double voting.

Another solution is to lock down those who can vote at the start of a motion, so that anyone who gets voice or joins after voting has started cannot vote. This allows the ops to know who can vote before there is a flood of ayes, nays or parts and joins. There is an issue if a user loses connection part way through a motion and cannot vote or change their vote.

I see some issues with tallying votes by host mask. Users on the same local network may have the same host mask and not be counted separately. It also leaves room for a mistake if a voter loses power and switches to a mobile network with a different host mask.

**Comment by DanielOaks** [comment]
The voicing is done via username+hostmask rather than nicks, so that's not as much of an option unfortunately. In that sort of situation where they rejoin with their original nick, the repeat vote would simply overwrite the original vote, because they're stored primarily by nick right now.

With tallying votes, I mean by username+hostmask as is done for recognised users, for instance: `dan@198.51.100.32`, which should make sure users on the same local network get counted differently. It does still leave room for the mobile network issue, but we could put a note in the readme about recommending keeping voting periods on motions relatively shortish to try and prevent stuff like that.

This username+hostmask for tallying votes sorta solution is more aimed to offset the situation where someone times out, their client rejoinins with a new temporary nick, they get revoiced by services and accidentally vote again under their new nick. That mobile net type of issue isn't really within our reach unless we get into weird account junk which is difficult and gets annoying to administer, unfortunately.

**Comment by olligobber** [comment]
As long as a user cannot leave the room and rejoin without manually being given voice by an op, it should avoid most confusion and vote gaming, regardless of whether voicing is done by mask or nick.

If you tally votes by username+hostmask it still allows users to change nick or hostname to get another vote. If you look for users that rejoin with the same username or hostmask as one who left, it leaves room for a different user on the same network or with the same nick to be confused with one who had already voted.

**Comment by DanielOaks** [comment]
Yep, it would avoid confusion and vote gaming. However, the `!add` command adds the user's user+host mask to its list of recognised users. With the original Motionbot (I'll make a PR to readd the functionality to Rhythm), if a user joins the channel and their user+host is in the recognised list, the bot will voice the newly-joined user. The voicing is designed to be automatically done, rather than something the ops need to do manually with each user joining. That's why I think looking at this can be useful, particularly for the accidental pingout+rejoin.

If we were doing that sort of checking, the user could change their user/host and get a new vote. However, there's not that much of a way for us to stop that since we're not network operators.

Given that we'd be checking user+host, we shouldn't get two different users showing up as the same user. The username is either a unique identifier for a user on an IP address for standard ident servers, or a user-set value that doesn't change between reconnections, unlike nicknames can (for instance, a user's nickname could be `dan` vs `dan_` if a user reconnects before their old nick has timed out, but they'd both still have the same user+host mask). Web clients seem to fall on the fence though, for instance mibbit users seem to just have the same username `mibbit`, but they pass the original IP address through. Very irritating set of issues.

Regarding the first point, maybe users joining during the voting period for a motion don't get automatically voiced until the motion ends? Just throwing ideas out right now, seeing what sticks and what doesn't.

**Comment by olligobber** [comment]
I'm not so sure the `!add` command should automatically voice users purely based on any combination of nick/user/hostmask as these can be spoofed. However, since (at the moment) the only authority in being voiced is being able to vote, disabling automatic voice during a motion would prevent any abuse of the bot, without manual intervention by an op.

I had not thought of using username to differentiate users. There are numerous clients with this issue, mibbit, quassel etc. I'm not sure how those handle realname though.

Also, consider that some users have multiple devices. For example
    [olligobber] (~olligobbe@255-215-7-149.static.tpgi.com.au): realname
    [sogvii] (~olligobbe@255-215-7-149.static.tpgi.com.au): olligobber
